src/marin/processing/classification/inference.py

```diff
@@ -26,215 +26,424 @@
+                pq.write_table(table, f)
+    else:
+        raise ValueError(f"Unsupported filetype: {output_filename}")
 
-    write_dataset(dataset, output_filename)
 
+def get_input_dataset_column_names(dataset_schema: DatasetSchemaConfig | None = None) -> list[str]:
+    schema = dataset_schema or DatasetSchemaConfig()
+    return schema.input_columns
 
-@ray.remote(max_calls=1)
-@remove_tpu_lockfile_on_exit
-def process_file_ray(
+
+def get_output_dataset_column_names(dataset_schema: DatasetSchemaConfig | None = None) -> list[str]:
+    schema = dataset_schema or DatasetSchemaConfig()
+    return schema.output_columns
+
+
+@cached_or_construct_output(success_suffix="SUCCESS")
+def process_file_with_quality_classifier_streaming(
     input_filename: str,
     output_filename: str,
+    # quality_classifier: BaseClassifier,
     model_name_or_path: str,
     attribute_name: str,
-    model_type: str | None,
-    filetype: str,
+    model_type: str,
     classifier_kwargs: dict,
+    dataset_schema: DatasetSchemaConfig,
+    batch_size: int = 512,
+    resume: bool = True,
+    use_autoscaling_actor_pool: bool = False,
+    classifier_actor_options: dict | None = None,
 ):
-    quality_classifier = AutoClassifier.from_model_path(
-        model_name_or_path, attribute_name, model_type, **classifier_kwargs
-    )
+    """Process a file with streaming I/O and resumption capability"""
+    print(f"[*] Processing {input_filename} to {output_filename}")
 
-    process_file_with_quality_classifier(input_filename, output_filename, quality_classifier)
+    # Initialize mutable defaults
+    if classifier_actor_options is None:
+        classifier_actor_options = {}
+
+    # Get column names
+    input_column_names = get_input_dataset_column_names(dataset_schema)
+    output_column_names = get_output_dataset_column_names(dataset_schema)
+
+    # Check if we should resume from existing progress by loading finished IDs
+    finished_ids = set()
+    if resume:
+        finished_ids = get_finished_ids(output_filename, dataset_schema.id_column)
+        if finished_ids:
+            print(f"[*] Resuming: found {len(finished_ids)} already processed IDs")
+        else:
+            print(f"[*] No existing IDs found in {output_filename}")
+
+    # Create streaming iterator
+    row_iterator = read_dataset_streaming(input_filename, input_column_names)
+
+    # Initialize for batch processing
+    append_mode = len(finished_ids) > 0
+
+    # For parquet, collect batches; for JSONL, stream asynchronously
+    # if output_filename.endswith(".parquet"):
+    #     parquet_batches = []
+    # async_writer = None
+    # else:
+    # async_writer = AsyncJSONLWriter(output_filename, append=append_mode)
+
+    batch = []
+    total_processed = len(finished_ids)
+    total_skipped = 0
+
+    task_queue = Queue()
+    result_queue = Queue()
+
+    if use_autoscaling_actor_pool:
+        pool = AutoscalingActorPool(
```

Comment:
e.g. here instead of having a bool for use_actor_pool, you could have an num_classifier_workers: int , and use the actor pool if num_classifier_workers > 1